pkg/image/index/index.go

```diff
@@ -0,0 +1,179 @@
+package index
+
+import (
+	"time"
+
+	"github.com/docker/distribution/digest"
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/client"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type ImageIndex struct {
+	client client.ImageInterface
+
+	imageController  *framework.Controller
+	imageIndex       cache.Indexer
+	imageStoreSynced func() bool
+}
+
+var (
+	ChainIndexName     = "chain"
+	LastChainIndexName = "last-chain"
+	// TODO: Re-listing images might be expensive on large clusters, this should be
+	// configurable?
+	ResyncInterval = 10 * time.Minute
+)
+
+// NewImageIndex returns a new image index.
+func NewImageIndex(client client.ImageInterface, stopChan <-chan struct{}) *ImageIndex {
+	c := &ImageIndex{
+		client: client,
+	}
+	c.imageIndex, c.imageController = framework.NewIndexerInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return c.client.List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return c.client.Watch(options)
+			},
+		},
+		&imageapi.Image{},
+		ResyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    c.addImage,
+			UpdateFunc: c.updateImage,
+			DeleteFunc: c.deleteImage,
+		},
+		cache.Indexers{
+			ChainIndexName:     chainIndexFunc,
+			LastChainIndexName: lastChainIndexFunc,
+		},
+	)
+
+	go c.imageController.Run(stopChan)
+	return c
+}
+
+func (c *ImageIndex) WaitForSyncedStores() {
+	for !c.imageController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// Descendants provides a list of images that the provided image is based on.
+// Note that if a base image is tagged multiple times, this will return all the matching
+// images.
+func (c *ImageIndex) Descendants(image *imageapi.Image) ([]*imageapi.Image, error) {
```

Comment:
@mfojtik that is highly confusing. The indexed data structure is a forest (a set of disjoinet trees). And according to [wikipedia](https://en.wikipedia.org/wiki/Tree_(data_structure)):

> Descendant - A node reachable by repeated proceeding from parent to child.
> Ancestor - A node reachable by repeated proceeding from child to parent.
> Child - A node directly connected to another node when moving away from the Root.
> Root - The top node in a tree.

Where the `Root` is a hash of the very first layer in an image.

Instead of violating a terminology here, would it be perhaps more feasible to name it like `ParentImages`?